vendor/random/random.cc

```diff
@@ -0,0 +1,116 @@
+/* Copyright (c) 2019 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/vendor/random/random.h"
+
+#include <cmath>
+
+#include "base/bits.h"
+#include "crypto/random.h"
+
+namespace brave {
+namespace random {
+
+uint64_t Uniform64() {
+  uint64_t x;
+
+  crypto::RandBytes(&x, sizeof x);
+
+  return x;
+}
+
+// Correct floating-point uniform [0,1] sampler which gives exactly
+// the correct weight to every floating-point number in [0,1],
+// i.e. the Lebesgue measure of its binade.
+//
+// In principle this algorithm could return any floating-point number
+// in [0,1], but in practice it cannot return 0 because the
+// probability is 2^-1075; consequently you can reliably pass the
+// result to, e.g., log, and be guaranteed to get a finite result.
+//
+// In contrast, if we simply divided a uniform random 53-bit or 64-bit
+// integer by 2^53 or 2^64, the result would _not_ be guaranteed to be
+// nonzero, _and_ it would exclude the result 1, which it should
+// return with probability 2^-54.
+double Uniform_01() {
+  uint64_t e, x, u;
+
+  // Draw an exponent with geometric distribution.
+  e = 0;
+  do {
+    if ((x = Uniform64()) != 0)
+      break;
+    e += 64;
+  } while (e < 1088);
+
+  // Count the remaining leading zero bits to finish up the geometric
+  // draw.
+  //
+  // If we stopped at e >= 1088, this means our RNG is broken.  In
+  // that case, we could just as well abort the process.  But it is
+  // also safe to call CountLeadingZeroBits at this point; it will
+  // just return 64, and the exponent will be even more improbably
+  // larger.
+  e += base::bits::CountLeadingZeroBits(x);
+
+  u = Uniform64();
+
+  return deterministic::Uniform_01(e, u);
+}
+
+// Nondeterministic distribution samplers.  These should call
+// Uniform64 and Uniform_01 only, and pass them on to a deterministic
+// transform in order to facilitate automatic testing.
+
+double Exponential(double rate) {
+  uint64_t s = Uniform64();
+  double p0 = Uniform_01();
+  return deterministic::Exponential(s, p0, rate);
+}
+
+uint64_t Geometric(double period) {
+  uint64_t s = Uniform64();
+  double p0 = Uniform_01();
+  return deterministic::Geometric(s, p0, period);
+}
+
+namespace deterministic {
```

Comment:
I think we should avoid nesting too many namespaces, especially when there are overlapping function names inside them. People will add `using` declarations and then it's easier to confuse which function is being called. The usual convention in chromium is just the module, but we're using `brave::` to avoid conflicts. Based on the comments in the header file I'm not sure if we actually need to expose these or not? Can they be in an anonymous namespace in the `cc` file or do we want these exposed in the header? If they should be exposed them maybe `DeterministicStdExponential`, etc... instead of `deterministic` namespace so there's no overlap in the function names?

Reply:
These are used in random.cc and in random_unittest.cc.  The overlap is intentional: many of the interesting _stochastic_ samplers in random.cc are very small functions in terms of (a) some uniform samples and then (b) a deterministic transformation into the desired distribution.

Testing stochastic samplers is annoying (stochastic distribution tests are qualitatively different from other kinds of automatic tests, necessarily having false alarm rates), but the deterministic transformations are easy to test.

Since most users will not need the deterministic ones, while both random.cc and random_unittest.cc need them, I put them into a separate namespace on purpose.

Reply:
I could introduce automatic tests for the stochastic functions, but then we have to start worrying about how to handle the unavoidable false alarms.

For a small change like this, I figured it would be much easier all around if

1. the stochastic parts of the new code as everything will actually use are mostly just wiring a uniform integer sampler to a deterministic function, small enough to confidently audit by reading, and
2. the deterministic parts as used internally in their own namespace are comfortably tested with deterministic tests.